### PR TITLE
signalr feature WIP

### PR DESCRIPTION
**Pull request by ianadavies**
Initial PR, NB: This is just to get some initial feedback as this is still a WIP.

Preliminary work required for SignalR implementation.

This PR improves how we handle observables in the components, brings in the use of the async pipe in the templates to manage subscriptions to those observables.

BehavourSubjects are now used in the services where applicable.

Staking and Transactions are now their own components to improve SRP and make the code cleaner and easier to follow. Staking service has been created to manage the staking subscriptions.

Base class for RestApi calls has also been created and an ErrorService that can be dependency injected.

The idea here is that we will slowly deprecate the ApiService, and this will be replaced with smaller services with a specific responsibility. For now I have created a WalletService and StakingService.

Open to suggestions on this.

Also please note : The following code has been removed from the polling subscriptions in the components as it didn't make much sense to me 

```
error => {
        if (error.status === 0) {
          this.cancelSubscriptions();
        } else if (error.status >= 400) {
          if (!error.error.errors[0].message) {
            this.cancelSubscriptions();
            this.startSubscriptions();
          }
        }
      }
```

I have spoken to Igor and Peterjan about this and it seems this is only due to bug in the Api. Now we use behaviourSubjects in our services we can now handle these error conditions there' instead as it shouldn't be responsibility of the components to do this.

If I can get all the events I need to broadcast from the node through SignalR then this error handling should not be needed anyway.

### CODE REVIEW

File: StratisCore.UI/src/app/login/login.component.ts [hunk 1]
```diff
@@ -37,14 +55,16 @@ export class LoginComponent implements OnInit {
       "password": [{value: "", disabled: this.isDecrypting}, Validators.required]
     });
 
-    this.openWalletForm.valueChanges
-      .subscribe(data => this.onValueChanged(data));
+    this.subscriptions.push(this.openWalletForm.valueChanges
```
Comment (fenix2222) [inline StratisCore.UI/src/app/login/login.component.ts]: See if we can manage subscriptions a bit better. Under: https://github.com/stratisproject/StratisCore/tree/master/StratisCore.UI/src/app/wallet/tokens/models we have `disposable` implementation. You can see how it is used here: https://github.com/stratisproject/StratisCore/blob/master/StratisCore.UI/src/app/wallet/tokens/components/tokens.component.ts (Mixin + implementing disposable)

File: StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+})
+export class SecondsToStringPipe implements PipeTransform {
+  constructor() {
+  }
+
+  public transform(seconds: number): string {
+    let numDays = Math.floor(seconds / 86400);
+    let numHours = Math.floor((seconds % 86400) / 3600);
+    let numMinutes = Math.floor(((seconds % 86400) % 3600) / 60);
+    let numSeconds = ((seconds % 86400) % 3600) % 60;
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts]: not used?

File: StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    if (numDays > 0) {
+      if (numDays > 1) {
+        dateString += numDays + " days ";
+      } else {
+        dateString += numDays + " day ";
+      }
+    }
+
+    if (numHours > 0) {
+      if (numHours > 1) {
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts]: could simplify nesting by using just 
```
dateString += `${numDays} day${numDays > 1 ? 's' : ''} `;
```

or
```
private toPlural(value: number) {
    return value > 1 ? 's' : '';
}

dateString += `${numDays} day${toPlural(numDays)} `;
```

and apply it in this class

File: StratisCore.UI/src/app/shared/services/interfaces/signalr-events.i.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { GeneralInfo, StakingInfo } from "@shared/services/interfaces/api.i";
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/services/interfaces/signalr-events.i.ts]: Not a big one, but we have a mixture of " and ', in most old components it is ' and TSLint prefers '

File: StratisCore.UI/src/app/shared/services/signalr-service.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  public registerOnMessageEventHandler<TMessage>(messageType: string, onEventMessageReceivedHandler: (message: TMessage) => void): void {
+    this.onMessageReceivedHandlers.push({
+      messageType: messageType,
+      onEventMessageDelegate: onEventMessageReceivedHandler
+    });
+  }
+
+  public onConnectionFailed: EventEmitter<Error> = new EventEmitter<Error>();
+
+  public connect(hubName: string): void {
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/services/signalr-service.ts]: Do we need any connection recovery logic, for example if connection drops for any reason, show we try to reconnect?

File: StratisCore.UI/src/app/shared/services/staking-service.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    this.invokeStartStakingApiCall(walletData)
+      .toPromise().then(
+      () => {
+        this.stakingEnabled = true;
+        this.isStarting = false;
+        return true;
+      },
+      error => {
+        console.log(error);
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/services/staking-service.ts]: Use `Log.debug()` or `Log.error` from logger.service, it does formatting and doesn't log for release mode.

File: StratisCore.UI/src/app/shared/services/signalr-service.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      this.connection.onclose((error: Error) => {
+        this.onConnectionFailed.emit(error);
+        console.log(`disconnected ${error.message}`);
+      });
+
+      console.log("connecting...");
+
+      this.connection.start()
+        .then(() => {
+          console.log("Connected!");
```
Comment (fenix2222) [inline StratisCore.UI/src/app/shared/services/signalr-service.ts]: Use `Log.debug()` or `Log.info` from logger.service, it does formatting and doesn't log for release mode. Check this entire file, it uses console.log a lot.

File: StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts (hunk 2 above)
Comment (ianadavies) [reply StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts]: It's used in the staking.component.html file

File: StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts (hunk 3 above)
Comment (ianadavies) [reply StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts]: This was existing code that I moved from dashboard.cshtml but yes I can improve

File: StratisCore.UI/src/app/shared/services/signalr-service.ts (hunk 5 above)
Comment (ianadavies) [reply StratisCore.UI/src/app/shared/services/signalr-service.ts]: I have add this yesterday

File: StratisCore.UI/src/app/login/login.component.ts (hunk 1 above)
Comment (ianadavies) [reply StratisCore.UI/src/app/login/login.component.ts]: If you look at the previous code here the subscription was left un-assigned and was not managed at all, this was a potential memory leak. I now add this the subscriptions array and on `ngOnDestroy()` I unsubscribe
```
public ngOnDestroy(): void {
    this.subscriptions.forEach(subscription => subscription.unsubscribe());
  }
```
I have looked at `tokens.component.ts` and read this article 
[https://blog.angularindepth.com/rxjs-avoiding-takeuntil-leaks-fb5182d047ef](https://blog.angularindepth.com/rxjs-avoiding-takeuntil-leaks-fb5182d047ef)

To be honest this code looks very confusing and problematic to me and I am don't really understand the benefit of using this takeUntil operator to manage subscriptions as it's not a common pattern and not obvious to the developer whats happening. You also have to remember to make sure takeUntil is the last operator or the developer will risk a memory leak still.

Angular natively supports management of subscriptions using the template `| async` pipe which I was going to apply to apply more of the code base as a separate exercise to this.

The `tokens.component.ts` also looks far too big IMHO and feels like it is mixing responsibility. I would expect a lot of the work being done in the component should be done in the `TokenService` or `SmartContractService`

Comment (fenix2222) [reply StratisCore.UI/src/app/login/login.component.ts]: There are lots of articles about this stuff, but the usual consensus is to use `| async` pipe where possible, if not possible use `takeUntil` and `subscriptions` as last. If you can use `| async` pipe that would be the best.

File: StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts (hunk 2 above)
Comment (fenix2222) [reply StratisCore.UI/src/app/shared/pipes/seconds-to-string.pipe.ts]: I am referring to `numSeconds` variable. It comes up as unused, how does html use a locally scoped variable?

**Comment by ArcSin2000X** [comment]
This PR requires SignalR feature from SBFN to work which is not present on current submodule which is 2 months old. I see 2 options forward: either back out the PR or update SBFN submodule to master. @fenix2222 @dev0tion

**Comment by dev0tion** [comment]
Thanks @ArcSin2000X, will update submodule to master.

**Pull request merged**